File: proj/v8seg/onnx/main.cpp
#include <iostream>
#include <fstream>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>
#include <onnxruntime_cxx_api.h>

class YOLOV8SEG
{
public:
	YOLOV8SEG(float confThreshold, float nmsThreshold, std::string modelpath);
	void drawPred(cv::Mat& frame, cv::Rect box, int classid, float conf);
	void normalize_(cv::Mat img);
	void maskimg2pred(cv::Mat& frame, cv::Mat maskpreds, std::vector<cv::Rect> npboxes, float* maskout);
	void detect(cv::Mat& frame);
private:

	int inpWidth;
	int inpHeight;
	int nm;
	int numproposal;
	int nout;


	float confThreshold;
	float nmsThreshold;

	int mask_width;
	int mask_height;

	std::vector<float> input_image_;
	std::vector<std::string> class_names;


	std::vector<const char*> input_names{"images"};
	std::vector<const char*> output_names{"output0", "output1"};

	Ort::Env env = Ort::Env(ORT_LOGGING_LEVEL_ERROR, "YOLOV8SEG");

	Ort::Session *sess = nullptr;
	Ort::SessionOptions sessops = Ort::SessionOptions();

};

YOLOV8SEG::YOLOV8SEG(float confThreshold, float nmsThreshold, std::string modelpath)
{
	this->confThreshold = confThreshold;
	this->nmsThreshold = nmsThreshold;

	this->inpWidth = 640;
	this->inpHeight = 640;
	this->nm = 32;
	this->numproposal = 8400;
	this->nout = 116;

	this->mask_width = 160;
	this->mask_height = 160;

	std::ifstream ifs("coco.names");
	std::string line;

	while(getline(ifs, line)) this->class_names.push_back(line.substr(0, line.length()-1));

	OrtSessionOptionsAppendExecutionProvider_CUDA(this->sessops, 0);

	this->sessops.SetGraphOptimizationLevel(ORT_ENABLE_BASIC);

	this->sess = new Ort::Session(this->env, modelpath.c_str(), this->sessops);

}

void YOLOV8SEG::drawPred(cv::Mat& frame, cv::Rect box, int classid, float conf)
{
	cv::rectangle(frame, box, cv::Scalar(0,144,255), 2);

	std::string label = cv::format("%.2f", conf);

	label = this->class_names[classid] + ":" + label;

	cv::putText(frame, label, cv::Point(box.x,box.y-10), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(144,0,255), 2);


}


void YOLOV8SEG::normalize_(cv::Mat img)
{
	int srch = img.rows, srcw = img.cols;
	this->input_image_.resize(srcw * srch * img.channels());

	for(int c = 0; c < 3; ++c)
	{
		for(int i = 0; i < srch; ++i)
		{
			for(int j = 0; j < srcw; ++j)
			{
				float pix = img.ptr<uchar>(i)[j * 3 + 2 - c];
				this->input_image_[c*srcw*srch+i*srcw+j] = pix / 255.0;
			
			}
		}
	
	}

}


void YOLOV8SEG::maskimg2pred(cv::Mat& frame, cv::Mat maskpreds, std::vector<cv::Rect> npboxes, float* maskout)
{
	cv::Mat protos = cv::Mat(this->nm, this->mask_width*this->mask_height, CV_32F, maskout);
	cv::Mat maskRes = (maskpreds * protos).t();
	cv::Mat masks = maskRes.reshape(npboxes.size(), {this->mask_height, this->mask_width});

	float mratiow = (float)frame.cols / this->mask_width;
	float mratioh = (float)frame.rows / this->mask_height;

	cv::Size blur_size = cv::Size(int(mratiow), int(mratioh));
	cv::Mat mask_img = frame.clone();

	cv::RNG rng(456);

	std::vector<cv::Mat> maskChannels;
	cv::split(masks, maskChannels);

	for(int i = 0; i < npboxes.size(); ++i)
	{
		int ox = npboxes[i].x, oy = npboxes[i].y, ow = npboxes[i].width, oh = npboxes[i].height;
		int sx = int(ox / mratiow), sy = int(oy / mratioh), sw = int(ow / mratiow), sh = int(oh / mratioh);

		cv::Mat dest;
		cv::exp(-maskChannels[i], dest);
		dest = 1 / (1 + dest);

		cv::Rect roi(sx, sy, sw, sh);
		cv::Mat scale_crop_mask = dest(roi);

		cv::Mat crop_mask;
		cv::resize(scale_crop_mask, crop_mask, cv::Size(ow, oh), cv::INTER_CUBIC);

		cv::blur(crop_mask, crop_mask, blur_size);

		crop_mask = crop_mask > 0.5;

		int b = rng.uniform(0, 255), g = rng.uniform(0, 255), r = rng.uniform(0, 255);
		mask_img(npboxes[i]).setTo(cv::Scalar(b,g,r), crop_mask);
	
	}

	cv::addWeighted(mask_img, 0.5, frame, 0.5, 0, frame);


}

void YOLOV8SEG::detect(cv::Mat& frame)
{
	cv::Mat dsimg;
	cv::resize(frame, dsimg, cv::Size(this->inpWidth, this->inpHeight));

	this->normalize_(dsimg);

	std::array<int64_t, 4> input_shape_{1, 3, this->inpHeight, this->inpWidth};

	auto allocator_info = Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeCPU);
	Ort::Value input_tensor = Ort::Value::CreateTensor<float>(allocator_info, input_image_.data(), input_image_.size(), input_shape_.data(), input_shape_.size());

	std::vector<Ort::Value> output_tensors = this->sess->Run(Ort::RunOptions{nullptr}, &input_names[0], &input_tensor, 1, output_names.data(), output_names.size());

	std::vector<int> ids;
	std::vector<float> confs;
	std::vector<cv::Mat> mpreds;
	std::vector<cv::Rect> boxes;

	float ratiow = (float)frame.cols / this->inpWidth;
	float ratioh = (float)frame.rows / this->inpHeight;

	float* pred = output_tensors[0].GetTensorMutableData<float>();

	cv::Mat rawData = cv::Mat(cv::Size(this->numproposal, this->nout), CV_32F, pred).t();

	float* pdata = (float*)rawData.data;

	for(int n = 0; n < this->numproposal; ++n)
	{
		float maxss = 0.0;
		int idp = 0;
		for(int k = 0; k < this->nout-4-this->nm; ++k)
		{
			if(pdata[k + 4] > maxss)
			{
				maxss = pdata[k + 4];
				idp = k;
			
			}
		
		}
		if(maxss >= this->confThreshold)
		{
		
			float cx = pdata[0] * ratiow;
			float cy = pdata[1] * ratioh;
			float w = pdata[2] * ratiow;
			float h = pdata[3] * ratioh;

			int left = int(cx - 0.5 * w);
			int top = int(cy - 0.5 * h);

			boxes.push_back(cv::Rect(left, top, int(w), int(h)));
			confs.push_back(maxss);
			ids.push_back(idp);
			mpreds.push_back(rawData.row(n).colRange(this->nout-this->nm, this->nout));
		}
		pdata += this->nout;
	
	}

	std::vector<int> indices;
	cv::dnn::NMSBoxes(boxes, confs, this->confThreshold, this->nmsThreshold, indices);
	
	std::vector<cv::Rect> npboxes;
	cv::Mat maskpreds;
	float* maskout = output_tensors[1].GetTensorMutableData<float>();

	for(int idx : indices)
	{
		mpreds[idx].convertTo(mpreds[idx], CV_32F);
		npboxes.push_back(boxes[idx]);
		maskpreds.push_back(mpreds[idx]);
	
	}
	this->maskimg2pred(frame, maskpreds, npboxes, maskout);

	for(int idx : indices) this->drawPred(frame, boxes[idx], ids[idx], confs[idx]);


}


int main()
{
	YOLOV8SEG net(0.7, 0.8, "weights/yolov8n-seg.onnx");

	cv::Mat srcimg = cv::imread("imgs/person.jpg");

	net.detect(srcimg);

	cv::imwrite("imgs/person_onnx_cpp.jpg", srcimg);

	return 0;
}

File: proj/v8seg/onnx/main.py
import numpy as np
import onnxruntime
import cv2
import random

def sigmoid(x):
    return 1 / (1 + np.exp(-x))


class YOLOV8SEG:

    def __init__(self, modelpath, conf_thres=0.7, nms_thres=0.8):
        self.confThreshold = conf_thres
        self.nmsThreshold = nms_thres

        self.class_names = [x.strip() for x in open("coco.names", "r").readlines()]

        self.inpWidth = 640
        self.inpHeight = 640

        self.nm = 32

        self.sess = onnxruntime.InferenceSession(modelpath, providers=["CUDAExecutionProvider", "CPUExecutionProvider"])

        self.innames = [x.name for x in self.sess.get_inputs()]
        self.ounames = [x.name for x in self.sess.get_outputs()]


    def drawPred(self, image, box, classid, conf):

        x,y,w,h = box

        cv2.rectangle(image, (x,y), (x+w,y+h), (255,144,0), 2)

        label = self.class_names[classid] + ":" + str(round(conf, 2))

        cv2.putText(image, label, (x,y-10), cv2.FONT_HERSHEY_SIMPLEX, 1, (0,255,144), 2)

    def maskimg2pred(self, image, maskpreds, npboxes, maskout):
        num_mask, mask_height, mask_width = maskout.shape

        masks = sigmoid(maskpreds @ maskout.reshape((num_mask, -1))).reshape((-1, mask_height, mask_width))

        mask_img = image.copy()
        scale_npboxes = npboxes.copy().astype(np.float32)

        mratiow, mratioh = image.shape[1] / mask_width, image.shape[0] / mask_height

        blur_size = (int(mratiow), int(mratioh))

        scale_npboxes[:, [0,2]] /= mratiow
        scale_npboxes[:, [1,3]] /= mratioh

        for i in range(len(npboxes)):
            ox,oy,ow,oh = npboxes[i].astype(int)
            sx,sy,sw,sh = scale_npboxes[i].astype(int)

            scale_crop_mask = masks[i][sy:(sy+sh), sx:(sx+sw)]

            crop_mask = cv2.resize(scale_crop_mask, (ow,oh), cv2.INTER_CUBIC)

            crop_mask = cv2.blur(crop_mask, blur_size)

            crop_mask = (crop_mask > 0.5).astype(np.uint8)[..., np.newaxis]

            crop_mask_img = mask_img[oy:(oy+oh), ox:(ox+ow)]

            mask_img[oy:(oy+oh), ox:(ox+ow)] = crop_mask_img * (1 - crop_mask) + tuple(random.sample(range(255), 3)) * crop_mask
        maskalpha = 0.5
        return cv2.addWeighted(mask_img, maskalpha, image, 1-maskalpha, 0)

    def detect(self, image):

        blob = cv2.dnn.blobFromImage(image, 1.0 / 255.0, (self.inpHeight, self.inpWidth), (0,0,0), True, False)

        outs = self.sess.run(self.ounames, {self.innames[0]:blob})

        ratiow, ratioh = image.shape[1] / self.inpWidth, image.shape[0] / self.inpHeight

        boxes, confs, ids, mpreds = [], [], [], []

        for pred in np.squeeze(outs[0]).transpose():
            maxss = np.max(pred[4:-self.nm])
            idp = np.argmax(pred[4:-self.nm])

            if maxss >= self.confThreshold:
                cx = pred[0] * ratiow
                cy = pred[1] * ratioh
                w = pred[2] * ratiow
                h = pred[3] * ratioh

                left = int(cx - 0.5 * w)
                top = int(cy - 0.5 * h)

                boxes.append([left, top, int(w), int(h)])
                confs.append(maxss)
                ids.append(idp)
                mpreds.append(list(pred[-self.nm:]))

        indices = cv2.dnn.NMSBoxes(boxes, confs, self.confThreshold, self.nmsThreshold)

        npboxes = np.array(boxes)[indices]
        maskpreds = np.array(mpreds)[indices]
        maskout = np.squeeze(outs[1])

        image = self.maskimg2pred(image, maskpreds, npboxes, maskout)

        for idx in indices:
            self.drawPred(image, boxes[idx], ids[idx], confs[idx])
        return image

if __name__ == "__main__":

    net = YOLOV8SEG("weights/yolov8n-seg.onnx")

    srcimg = cv2.imread("imgs/person.jpg")

    tarimg = net.detect(srcimg)

    cv2.imwrite("imgs/person_onnx_py.jpg", tarimg)
